### ISSUE TITLE
uki installed system starts interactive-installer after first boot

### ISSUE THREAD

**Issue by jimmykarily** [kairos-io/kairos#2773]
Reproduce:

- Create a UKI iso with enki
- Boot it
- Complete the interactive installer to install the system
- Reboot

What happens:

Momentarily in tty1 the interactive installer is shown again. After a while it's replaced by a login prompt. The `kairos-agent state get boot` command says `active_boot`.

What was expected:

The interactive installer shouldn't be started at all

`/etc/os-release`:

```
kairos@localhost:~$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
KAIROS_FLAVOR="ubuntu"
KAIROS_RELEASE="v3.1.1-9-g496a7fad"
KAIROS_GITHUB_REPO="kairos-io/kairos"
KAIROS_SOFTWARE_VERSION_PREFIX="k3s"
KAIROS_ID_LIKE="kairos-core-ubuntu-22.04"
KAIROS_VERSION_ID="v3.1.1-9-g496a7fad"
KAIROS_PRETTY_NAME="kairos-core-ubuntu-22.04 v3.1.1-9-g496a7fad"
KAIROS_VARIANT="core"
KAIROS_VERSION="v3.1.1-9-g496a7fad"
KAIROS_ARTIFACT="kairos-ubuntu-22.04-core-amd64-generic-v3.1.1-9-g496a7fad"
KAIROS_FAMILY="ubuntu"
KAIROS_HOME_URL="https://github.com/kairos-io/kairos"
KAIROS_NAME="kairos-core-ubuntu-22.04"
KAIROS_MODEL="generic"
KAIROS_BUG_REPORT_URL="https://github.com/kairos-io/kairos/issues"
KAIROS_FLAVOR_RELEASE="22.04"
KAIROS_TARGETARCH="amd64"
KAIROS_REGISTRY_AND_ORG="quay.io/kairos"
KAIROS_ID="kairos"
KAIROS_IMAGE_REPO="quay.io/kairos/ubuntu:22.04-core-amd64-generic-v3.1.1-9-g496a7fad"
KAIROS_IMAGE_LABEL="22.04-core-amd64-generic-v3.1.1-9-g496a7fad"
```

**Comment by jimmykarily** [comment]
Turns out various services are enabled when they shouldn't:

```
root@localhost:/home/kairos# systemctl status kairos-interactive
● kairos-interactive.service - kairos interactive-installer
     Loaded: loaded (/etc/systemd/system/kairos-interactive.service; enabled; vendor preset: enabled)
     Active: activating (start-pre) since Mon 2024-07-29 10:58:28 UTC; 26min ago
Cntrl PID: 1315 ((sh))
      Tasks: 1 (limit: 6178)
     Memory: 4.0K
        CPU: 77ms
     CGroup: /system.slice/kairos-interactive.service
             └─1315 "(sh)"

Jul 29 10:58:28 localhost systemd[1]: Starting kairos interactive-installer...

root@localhost:/home/kairos# ls /etc/systemd/system/multi-user.target.wants/ | grep interactive
kairos-interactive.service

root@localhost:/home/kairos# ls /etc/systemd/system/multi-user.target.wants/ | grep recovery
kairos-recovery.service
```

Might be this section in the unit file:

```
[Install]
WantedBy=multi-user.target
```

Maybe we never noticed [before we added](https://github.com/kairos-io/packages/commit/176010973c918f92320da89b82df6b135803bde8#diff-153837f65fee672b3f1558525fcf5e5a954f777c4bd08e96ffb623c4a9aacb3dR3):

```
After=multi-user.target
```

**Comment by Itxaka** [comment]
I cant reproduce this

```
root@localhost:~# systemctl status kairos-interactive
○ kairos-interactive.service - kairos interactive-installer
     Loaded: loaded (/etc/systemd/system/kairos-interactive.service; static)
     Active: inactive (dead)
root@localhost:~# s /etc/systemd/system/multi-user.target.wants/ | grep interactive
-bash: s: command not found
root@localhost:~# systemctl status kairos-interactive
○ kairos-interactive.service - kairos interactive-installer
     Loaded: loaded (/etc/systemd/system/kairos-interactive.service; static)
     Active: inactive (dead)
root@localhost:~# ls /etc/systemd/system/multi-user.target.wants/ | grep interactive
root@localhost:~# ls /etc/systemd/system/multi-user.target.wants/ | grep recovery
root@localhost:~# cat /etc/os-release 
PRETTY_NAME="Ubuntu 24.04 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo
KAIROS_NAME="kairos-core-ubuntu-24.04"
KAIROS_ARTIFACT="kairos-ubuntu-24.04-core-amd64-generic-v3.1.1-12-g6db220d-dirty"
KAIROS_FAMILY="ubuntu"
KAIROS_VARIANT="core"
KAIROS_MODEL="generic"
KAIROS_GITHUB_REPO="kairos-io/kairos"
KAIROS_REGISTRY_AND_ORG="quay.io/kairos"
KAIROS_BUG_REPORT_URL="https://github.com/kairos-io/kairos/issues"
KAIROS_SOFTWARE_VERSION_PREFIX="k3s"
KAIROS_ID_LIKE="kairos-core-ubuntu-24.04"
KAIROS_VERSION_ID="v3.1.1-12-g6db220d-dirty"
KAIROS_IMAGE_LABEL="24.04-core-amd64-generic-v3.1.1-12-g6db220d-dirty"
KAIROS_RELEASE="v3.1.1-12-g6db220d-dirty"
KAIROS_HOME_URL="https://github.com/kairos-io/kairos"
KAIROS_ID="kairos"
KAIROS_VERSION="v3.1.1-12-g6db220d-dirty"
KAIROS_PRETTY_NAME="kairos-core-ubuntu-24.04 v3.1.1-12-g6db220d-dirty"
KAIROS_IMAGE_REPO="quay.io/kairos/ubuntu:24.04-core-amd64-generic-v3.1.1-12-g6db220d-dirty"
KAIROS_FLAVOR="ubuntu"
KAIROS_FLAVOR_RELEASE="24.04"
KAIROS_TARGETARCH="amd64"
```

**Comment by Itxaka** [comment]
cant reproduce on 22.04+uki+core:

![image](https://github.com/user-attachments/assets/9c23fdf4-de78-4dff-8e34-011a64028998)


I would say lets close this unless its reproducible somehow or we get the same issue reported. Thoughts @jimmykarily ?

**Comment by jimmykarily** [comment]
I have the original artifact (`quay.io/kairos/ubuntu:22.04-core-amd64-generic-v3.1.1-9-g496a7fad`) from the description here and it's always reproducible. I don't know what commit `496a7fad` this is though, it must have been a branch. Let me see if I can reproduce it with a new build from master. If not, we can close.

**Comment by Itxaka** [comment]
> I have the original artifact (`quay.io/kairos/ubuntu:22.04-core-amd64-generic-v3.1.1-9-g496a7fad`) from the description here and it's always reproducible. I don't know what commit `496a7fad` this is though, it must have been a branch. Let me see if I can reproduce it with a new build from master. If not, we can close.

can you push it to ttl maybe?

**Comment by jimmykarily** [comment]
I tried these steps from master and I can't reproduce:

```bash
~/workspace/kairos/kairos (master)*$ git rev-parse HEAD
0f314294d8a0d32dfcd2e9263a6db3f363b3385d
~/workspace/kairos/kairos (master)*$ earthly -P +base-image --BASE_IMAGE=ubuntu:22.04 --FAMILY=ubuntu --FLAVOR=ubuntu --FLAVOR_RELEASE=22.04 --MODEL=generic --VARIANT=core
...
~/workspace/kairos/kairos (master)*$ sudo luet util unpack --local quay.io/kairos/ubuntu:22.04-core-amd64-generic-v3.1.1-18-g0f314294 unpacked
...
~/workspace/kairos/kairos (master)*$ docker run --rm -v $PWD/unpacked:/unpacked -v $PWD/build:/result -v /home/dimitris/workspace/kairos/enki/keys/:/keys quay.io/kairos/osbuilder-tools:latest  build-uki dir:/unpacked --output-dir /result --keys /keys --output-type iso --boot-branding "Kairos UKI"
...
# boot the artifact with qemu (secure boot)
```

Looking at my `docker images` , I think I've already pushed the artifact as `ttl.sh/kairos_uki_v3.1.1-9-g496a7fad` but let's ignore this for now. I don't even remember which branch this was created from so maybe it was something with the changes in that branch.

I'll close now and re-open if this occurs again.

**Comment by Itxaka** [comment]
Seems like the trigger for this is the agent from master as I hit that today twice in different distros!

build the image with  `--KAIROS_AGENT_DEV_BRANCH=master` and it should hit.

Reopen!

**Comment by Itxaka** [comment]
also happens on non-uki!

**Comment by Itxaka** [comment]
kairos-reset IS triggering and that wasnt changed since 8 months ago. It makes no sense that it would be the change to the services in this case.

Maybe some other change??? and why would it trigger with newer agent only...

gonna start from scratch

**Comment by Itxaka** [comment]
Iso built with

```
earthly +iso --FLAVOR=ubuntu --FLAVOR_RELEASE=24.04 --FAMILY=ubuntu --MODEL=generic --VARIANT=core --BASE_IMAGE=ubuntu:24.04  
```

From master, commit `5b406bb5e1d9358494796980c6a81043ad2c006e`
Agent version 2.13.1
Immucore version 0.4.1
Overlay files 1.1.42
Boot from livecd, select manual install.
kairos-recovery is disabled
kairos-reset is disabled
Only sentinel /run/cos/live_mode exists
All looks good

**Comment by Itxaka** [comment]
Iso built with

```
earthly +iso --FLAVOR=ubuntu --FLAVOR_RELEASE=24.04 --FAMILY=ubuntu --MODEL=generic --VARIANT=core --BASE_IMAGE=ubuntu:24.04  --KAIROS_AGENT_DEV_BRANCH=v2.13.2
```

From master, commit `5b406bb5e1d9358494796980c6a81043ad2c006e`
Agent version 2.13.2
Immucore version 0.4.1
Overlay files 1.1.42
Boot from livecd, select manual install.
kairos-recovery is ENABLED
kairos-reset is ENABLED
kairos-interactive is ENABLED
Only sentinel /run/cos/live_mode exists

Looks terrible.


So the difference is agent 2.13.1 vs 2.13.2

This is the diff between them:
https://github.com/kairos-io/kairos-agent/compare/v2.13.1...v2.13.2

sdk was bumped from v0.2.3 to v0.3.2 which could maybe explain this?
yip was NOT updated, so we can discard a change in yip breaking this.
There is really not any code changes that could explain this other than the sdk.

**Comment by Itxaka** [comment]
Changes between sdk:
https://github.com/kairos-io/kairos-sdk/compare/v0.2.3...v0.3.2

Not a lot either :sob: dont see anything that can be triggering this...

**Comment by Itxaka** [comment]
this is even weirder, becuase the installer stuff is done on initramfs which is part of immucore. A newer agent should NOT affect this??????

**Comment by Itxaka** [comment]
in fact during initramfs, the service is NOT enabled:

```
2024-08-01T11:36:56Z DBG Reading '/system/oem/52_installer.yaml'
2024-08-01T11:36:56Z DBG Executing stage 'Start installer on tty1..2'
2024-08-01T11:36:56Z DBG {"sysinfo":{"version":"1.1.0","timestamp":"2024-08-01T11:36:52.813316689Z"},"node":{"hostname":"localhost","machineid":"6c911499a0224ea6abba69d0755cbb00","hypervisor":"kvm"},"os":{"name":"Ubuntu 24.04 LTS dracut-060+5-1ubuntu3.2 (Initramfs)","vendor":"ubuntu","version":"24.04","release":"24.04","architecture":"amd64"},"kernel":{"release":"6.8.0-39-generic","version":"#39-Ubuntu SMP PREEMPT_DYNAMIC Fri Jul  5 21:49:14 UTC 2024","architecture":"x86_64"},"product":{"name":"Standard PC (Q35 + ICH9, 2009)","vendor":"QEMU","version":"pc-q35-9.0","uuid":"6c911499-a022-4ea6-abba-69d0755cbb00"},"board":{},"chassis":{"type":1,"vendor":"QEMU","version":"pc-q35-9.0"},"bios":{"vendor":"EFI Development Kit II / OVMF","version":"0.0.0","date":"02/06/2015"},"cpu":{"vendor":"GenuineIntel","model":"12th Gen Intel(R) Core(TM) i9-12900K","speed":2000,"cache":16384,"cpus":8,"cores":8,"threads":8},"memory":{"type":"RAM","size":16000},"storage":[{"name":"vda","driver":"virtio_blk","size":42}],"network":[{"name":"enp1s0","driver":"virtio_net","macaddress":"52:54:00:e2:f0:6e"}]}
2024-08-01T11:36:56Z DBG did not find any device in layout
2024-08-01T11:36:56Z WRN (conditional) Skip 'Skipping stage (if statement error: failed to run (grep -q "interactive-install" /proc/cmdline || grep -q "install-mode-interactive" /proc/cmdline) && \
([ -f /run/cos/live_mode ] || [ -f /run/cos/uki_install_mode ]) && \
( [ -e "/sbin/systemctl" ] || [ -e "/usr/bin/systemctl" ] || [ -e "/usr/sbin/systemctl" ] || [ -e "/usr/bin/systemctl" ] )
: exit status 1)' stage name: 
```

**Comment by Itxaka** [comment]
base image from where the iso is built also doesnt have the services in wanted so they should not start by default:
```
oot@be387d17f1be:/# ls /etc/systemd/system/multi-user.target.wants/
console-setup.service         nfs-client.target         thermald.service
e2scrub_reap.service          open-vm-tools.service     tpm2-abrmd.service
grub-common.service           remote-fs.target          ua-reboot-cmds.service
grub-initrd-fallback.service  rpcbind.service           ubuntu-advantage.service
lldpd.service                 snmpd.service             zfs.target
machines.target               ssh.service
networkd-dispatcher.service   systemd-networkd.service
```

**Comment by Itxaka** [comment]
I did a dissect in all kairos version commits and it started on this commit: https://github.com/kairos-io/kairos-agent/commit/e3d7d3a0578595647c9f856f289b419c8b841d55

Which is the one that bumps the sdk among other things, so it MUST be an sdk thing

**Comment by Itxaka** [comment]
ok, they way this can be affected during initramfs is because the dracut module for immucore brings the kairos-agent into initramfs, so a broken agent in rootfs can affect the initramfs. Now I wonder why do we bring the agent and where its used during initramfs

 - rootfs and initramfs are run directly from immucore so agent is not involved there
 - as far as I know, nothing should be calling those services enablement during initramfs stages. And on boot/network/fs stages those services should nto be touched as they are all set to be on the initramfs stage
 - network agent COULD trigger the rootfs and initramfs stages but only if there is userdata but there is none in my tests and the logs show that they are not triggered

**Comment by Itxaka** [comment]
We found the issue.

We used to empty the /etc/machine-id file
Then sysinfo got updated and sneakily filled the machine-id to a random number, which broke our workflow
We sent a patch upstream and then workaround this by removing the file completely
But it seems that sysinfo was still filling it up in the background somehow, at least during building the rootfs
Now sysinfo accepted our patch and we updated to the latest fixed version
But as there is now nothing creating the machine-id file, systemd show that there was no file and created it and reloaded the whole services from /etc
This resulted in services that are under /etc/systemd/system to be read and "installed"
Because our installer services had a WANTEDBY stanza, they automatically got installed as part of the multi-user target and started on each boot

The solution is twofold.

 - Disable the Install section on our services. we either manually enable and link them to a target or we start the services manually. We dont need systemd to start them itself
 - Empty the machine-id file instead of removing it. This is a feature of systemd that tis a bit hidden. If there is no file, it does the first boot thing fully. While if there is a machine-id but its empty, it tries to generate one but thats all it does.
